type-definitions/most.d.ts

```diff
@@ -1,385 +1,788 @@
-export function skipUntil<A>(signal: Stream<any>, s: Stream<A>): Stream<A>;
-export function during<A>(timeWindow: Stream<Stream<any>>, s: Stream<A>): Stream<A>;
-export function throttle<A>(period: number, s: Stream<A>): Stream<A>;
-export function debounce<A>(period: number, s: Stream<A>): Stream<A>;
-
-export function timestamp<A>(s: Stream<A>): Stream<TimeValue<A>>;
-export function delay<A>(dt: number, s: Stream<A>): Stream<A>;
-
-export function fromPromise<A>(p: Promise<A>): Stream<A>;
-export function await<A>(s: Stream<Promise<A>>): Stream<A>;
-
-export function sample<A, B, R>(
-  fn: (a: A, b: B) => R,
-  sampler: Stream<any>,
-  a: Stream<A>,
-  b: Stream<B>
-): Stream<R>;
-export function sample<A, B, C, R>(
-  fn: (a: A, b: B, c: C) => R,
-  sampler: Stream<any>,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>
-): Stream<R>;
-export function sample<A, B, C, D, R>(
-  fn: (a: A, b: B, c: C, d: D) => R,
-  sampler: Stream<any>,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>
-): Stream<R>;
-export function sample<A, B, C, D, E, R>(
-  fn: (a: A, b: B, c: C, d: D, e: E) => R,
-  sampler: Stream<any>,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>,
-  e: Stream<E>
-): Stream<R>;
-
-export function sampleWith<A>(sampler: Stream<any>, s: Stream<A>): Stream<A>;
-
-export function zip<A, B, R>(
-  fn: (a: A, b: B) => R,
-  a: Stream<A>,
-  b: Stream<B>
-): Stream<R>;
-export function zip<A, B, C, R>(
-  fn: (a: A, b: B, c: C) => R,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>
-): Stream<R>;
-export function zip<A, B, C, D, R>(
-  fn: (a: A, b: B, c: C, d: D) => R,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>
-): Stream<R>;
-export function zip<A, B, C, D, E, R>(
-  fn: (a: A, b: B, c: C, d: D, e: E) => R,
-  a: Stream<A>,
-  b: Stream<B>,
-  c: Stream<C>,
-  d: Stream<D>,
-  e: Stream<E>
-): Stream<R>;
-
-export function recoverWith<A, B>(p: (a: B) => A, s: Stream<A>): Stream<A>;
-export function throwError(e: Error): Stream<any>;
-
-export function multicast<A>(s: Stream<A>): Stream<A>;
+declare module "most" {
+	type SeedValue<S, V> = { seed: S, value: V };
+	type TimeValue<V>    = { time: number, value: V };
+
+	// es5 does not have typings for this
+	// use any to allow type-casting
+	type Generator<A, B, C> = any
+	type Iterable = any;
+
+	type CreateGenerator<A> = (...args: Array<any>) => Generator<A|Promise<A>, any, any>;
+
+	export interface Sink<A> {
+		event(time: number, value: A): void;
+		end(time: number, value?: A): void;
+		error(time: number, err: Error): void;
+	}
+
+	export interface Task {
+		run(time: number): void;
+		error(time: number, e: Error): void;
+		dispose(): void;
+	}
+
+	export interface ScheduledTask {
+		task: Task;
+		run(): void;
+		error(err: Error): void;
+		dispose(): void;
+	}
+
+	export interface Timer {
+		now(): number;
+		setTimer(f: () => any, delayTime: number): Task | number;
```

Comment:
Hmm, I think we'll need to look at where Asap instances are used.  I don't think Asap needs to conform to Task or ScheduleTask.  It's mostly used as an opaque "handle" to an asap task used internally by ClockTimer only.  For example, the only thing Scheduler does with Asap instances is: 1) create them by calling timer.setTimer, 2) destroy them by passing them to timer.clearTimer, and 3) compare them to `null`.